Light editor column registration

ericmehl opened the pull request:
This adds the ability to register a column in the Light Editor for editing a parameter anywhere within a light's shader network. Previously only parameters on the light itself were editable.

Fixes #5561.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [ ] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

ericmehl commented:
I'm marking this as draft for now - the ability to edit the original nodes of a light's shader network is still a work in progress. I think I have enough functionality there to talk through my approach and validate it or not. But there are a number of edge cases still to be addressed like handling spline plug connections, component connections, etc.

Also note that the first two commits are from https://github.com/GafferHQ/gaffer/pull/5846 and will be dropped when that is merged.

ericmehl commented:
After talking through it with @johnhaddon, we decided to use the existing `NetworkBuilder` in `GafferScene::Shader` instead of duplicating what was turning out to be quite a bit of functionality. Instead, I've added a new `parameterSource()` method to both `GafferScene::ShaderPlug` and `GafferScene::Shader` (protected in the latter). This queries the `NetworkBuilder::m_shaders` map to see what shader node generated a particular shader handle.

I think I've gotten the gist of what we talked about right, and the tests I added for the initial round are passing and interactive tests are working.

I didn't do anything with the concept of adding syntax to tweak plugs to allow something like `input(surfaceShader.color).tint` in order to tweak parameters where you know the relationship to a shader handle but not necessarily the target shader handle itself.

But I think it's ready for a new look, especially for the API that I've added in https://github.com/GafferHQ/gaffer/pull/5857/commits/abbfde215e51e7f589d9656d36054ee9c51e2fd2

ericmehl commented on src/GafferSceneUI/ParameterInspector.cpp:
```diff
@@ -108,11 +108,22 @@ Gaffer::ValuePlugPtr ParameterInspector::source( const GafferScene::SceneAlgo::H
+		{
+			if( const auto shaderPlug = runTimeCast<ShaderPlug>( plug ) )
+			{
+				return const_cast<ValuePlug *>( shaderPlug->parameterSource( m_parameter ) );
```
This is my main concern for the new API - is this `const_cast` a valid solution? The `NetworkBuilder` works with `const` plugs throughout, which does seem right considering what `NetworkBuilder` is meant to do.

Just to see how it would go, I tried making the return plug non-`const`, but the ripple effects of that went all the way up to the `Shader` API. Which doesn't seem like a good thing.

johnhaddon replied:
I think a `const_cast` is a valid solution, but I don't think this is the place for it. You're right that NetworkBuilder should continue to work with `const` plugs, because it must never modify the graph. And it's also right that `ShaderPlug::parameterSource() const` should return a const result (if the caller only has const access to ShaderPlug, they should only be able to use it to get const access to the rest of the graph). But here we have non-const access to ShaderPlug, so ShaderPlug should also provide us with a non-const overload for `parameterSource()` that returns us a non-const plug.

johnhaddon commented on src/GafferSceneUI/ParameterInspector.cpp:
```diff
@@ -157,8 +157,12 @@ Gaffer::ValuePlugPtr ParameterInspector::source( const GafferScene::SceneAlgo::H
 		{
-			/// \todo Consider shader as well as name
-			if( tweak->namePlug()->getValue() == m_parameter.name.string() && tweak->enabledPlug()->getValue() )
+			const std::string tweakName = (
```
This can be lifted out of the loop, rather than repeated for every iteration.

johnhaddon commented on include/GafferScene/Shader.h:
```diff
@@ -158,6 +159,11 @@ class GAFFERSCENE_API Shader : public Gaffer::ComputeNode
 		/// cases.
 		virtual IECore::DataPtr parameterValue( const Gaffer::Plug *parameterPlug ) const;
 
+		const Gaffer::ValuePlug *parameterSource(
```
Should be private. The other protected stuff is all for derived classes to override, but we're adding something with the intention that only the ShaderPlug can get to it (via friendship).

johnhaddon commented on include/GafferScene/ShaderPlug.h:
```diff
@@ -69,6 +72,8 @@ class GAFFERSCENE_API ShaderPlug : public Gaffer::Plug
 		/// generated by the input shader.
 		IECore::ConstCompoundObjectPtr attributes() const;
 
+		const Gaffer::ValuePlug *parameterSource( const IECoreScene::ShaderNetwork::Parameter &parameter) const;
```
Needs docs - maybe something like this?

> /// Returns the plug responsible for authoring the specified `parameter` within a shader network obtained by a call to `attributes()`.

johnhaddon commented on python/GafferSceneUITest/ParameterInspectorTest.py:
```diff
@@ -59,12 +59,12 @@ def testName( self ) :
 
 	@staticmethod
-	def __inspect( scene, path, parameter, editScope=None, attribute="light" ) :
+	def __inspect( scene, path, parameter, editScope=None, attribute="light", shader = "" ) :
```
`parameter` and `shader` are really confusing as separate arguments, especially when they're separated and one is optional. What we're really specifying here is a single `ShaderNetwork.Parameter()` which stores both the shader handle and the parameter handle. I suggest we drop the `shader` parameter and just pass a complete `Parameter`to the `parameter` argument (falling back to the current setup if a string is passed instead).

johnhaddon commented on python/GafferSceneUITest/ParameterInspectorTest.py:
```diff
@@ -878,6 +878,125 @@ def testShaderAndParameter( self ) :
 			edit = addBTweak
 		)
 
+	def testSingleConnection( self ) :
```
These new tests seem like they wouldn't work without the next commit - perhaps they belong with it?

johnhaddon commented on python/GafferSceneUITest/ParameterInspectorTest.py:
```diff
@@ -878,6 +878,125 @@ def testShaderAndParameter( self ) :
+			edit = s["add"]["parameters"]["a"]
+		)
+
+	# \todo :
+	# - Test component connections
+	# - Test splines
+	# - Test disabled node passthrough
+	# - Detect cycles
```
I think we can forego more exhaustive testing now that we know the internal implementation is using NetworkBuilder.

johnhaddon commented on src/GafferScene/Shader.cpp:
```diff
@@ -213,6 +213,18 @@ class Shader::NetworkBuilder
+			{
+				if( handleAndHash.handle == parameter.shader )
+				{
+					return shader->parametersPlug()->getChild<ValuePlug>( parameter.name );
```
I think we want `descendant()` here rather than `getChild()`, so we can get parameters inside structs.

johnhaddon commented on src/GafferScene/ShaderPlug.cpp:
```diff
@@ -248,6 +248,22 @@ IECore::ConstCompoundObjectPtr ShaderPlug::attributes() const
+		{
+			if( parameter.shader.string().empty() )
+			{
+				return s->parametersPlug()->getChild<ValuePlug>( parameter.name );
```
I think we want `descendant()` here too.

johnhaddon commented on src/GafferSceneUI/ParameterInspector.cpp:
```diff
@@ -108,11 +108,22 @@ Gaffer::ValuePlugPtr ParameterInspector::source( const GafferScene::SceneAlgo::H
+				return const_cast<ValuePlug *>( shaderPlug->parameterSource( m_parameter ) );
+			}
 		}
-		return light->parametersPlug()->getChild<ValuePlug>( m_parameter.name );
```
Until Lights are guaranteed to provide an internal Shader, we should probably keep this line. That also has the benefit to the reader that we know we're exiting here, and we don't need to scroll down and down to see what's at the end of the function.

johnhaddon commented on src/GafferSceneUI/ParameterInspector.cpp:
```diff
@@ -108,11 +108,22 @@ Gaffer::ValuePlugPtr ParameterInspector::source( const GafferScene::SceneAlgo::H
+		{
+			if( const auto shaderPlug = runTimeCast<ShaderPlug>( plug ) )
+			{
+				return const_cast<ValuePlug *>( shaderPlug->parameterSource( m_parameter ) );
```
We can use `parameterSource()` in the ShaderAssignment case below as well - it's even easier there, and it means we can drop the `Switch` handling shenanigans because that is dealt with for us in `parameterSource()` now.

johnhaddon commented on Changes.md:
```diff
@@ -27,7 +27,7 @@ Improvements
   - Changed all the Cycles items in the Gaffer tab menu to title case.
 - PythonCommand : Added standard popup menu to `command` widget in NodeEditor.
-- LightEditor : Added the ability to register columns for editing any shader in the light's shader network with an Edit Scope or a `ShaderTweaks` node. The parameter to edit is registered in the form `shaderName.parameterName` when calling `GafferSceneUI.LightEditor.registerParameter`.
+- LightEditor : Added the ability to register columns for editing any shader in the light's shader network. The parameter to edit is registered in the form `shaderName.parameterName` when calling `GafferSceneUI.LightEditor.registerParameter`.
```
I think the fact that every commit has needed to edit this line suggests that there might be a more logical commit order now we know where we're heading :

1. Add `ShaderPlug::parameterSource()`.
2. Use `parameterSource()` in ParameterInspector.
3. Expose feature in LightEditor.

johnhaddon reviewed (commented):
Thanks Eric - the new direction on this is feeling much better. Detailed comments inline - I think we're close enough that you can just squash and rebase away as you address them...
Cheers...
John

ericmehl commented on python/GafferSceneUI/LightEditor.py:
```diff
@@ -146,7 +146,8 @@ def registerParameter( cls, rendererKey, parameter, section = None, columnName =
 		# the latter isn't hashable and we need to use it as a dict key.
 		if isinstance( parameter, str ) :
-			parameter = ( "", parameter )
+			shader, dot, param = parameter.rpartition( "." )
```
I initially went with `rpartition()` because `"parameter".partition(".") = ( "parameter", "", "" )`, so it would look like the shader name is `parameter`. I think the most correct solution is `partition()` but only if there's a `.` found in the string?

ericmehl commented:
I addressed all the notes above, thanks for all the improvements @johnhaddon. I didn't mark them "fixed" individually because I also squashed them into the original commits as I went, but I can note them if that makes the review easier.

The commit sequence is hopefully easier to follow now, and I reworked the tests to put the majority of new tests in a single test in `ParameterInspectorTest` in https://github.com/GafferHQ/gaffer/pull/5857/commits/880f828fed7799812206f850df2db8ec77af4bb7. I changed that to test using shaders instead of a light to keep it in the relevant commit.

Marking as ready for (another) review!

johnhaddon commented:
Thanks Eric!
